File: LeetCode/1121. Divide Array Into Increasing Sequences/solution.py
from collections import Counter
class Solution:
    def canDivideIntoSubsequences(self, nums: List[int], K: int) -> bool:
        cc = Counter(nums)
        # print(cc)
        return K*max(cc.values()) <= len(nums)


# Think in the frequency of the numbers and how this affects the number of sequences needed.

# What is the minimum number of sequences we need to form? Considering frequency of the numbers.

# Think about the least number of sequences to maximize the lengths.

# The number of sequences needed is equal to the maximum frequency of an element.

# How to put the other elements into sequences ? Think in a greedy approach.

File: LeetCode/1121. Divide Array Into Increasing Sequences/solution.cpp
// const int maxn = 1e5+5;
// int cnt[maxn];

// class Solution {
// public:
//     bool canDivideIntoSubsequences(vector<int>& nums, int k) {
//         memset(cnt, 0, maxn*sizeof(int));
//         for (auto x : nums)
//             cnt[x]++;
//         int p = 0;
//         for (int i = 1; i < maxn; i++)
//             p = max(p, cnt[i]);
//         return p*k <= nums.size();
//     }
// };

class Solution {
public:
    bool canDivideIntoSubsequences(vector<int>& nums, int k) {
        int res = 0, cnt = 0, lk = nums[0];
        for (int k : nums) {
            if (k == lk) ++cnt;
            else res = max(res, cnt), cnt = 1, lk = k;
        }
        res = max(res, cnt);
        return res * k <= nums.size();
    }
};